jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java

```diff
@@ -0,0 +1,415 @@
+    private ByteBuffer _empty;
+    private ByteBuffer _inflated;
+
+    public GZIPContentDecoder()
+    {
+        this(null,2048);
+    }
+
+    public GZIPContentDecoder(int bufferSize)
+    {
+        this(null,bufferSize);
+    }
+    
+    public GZIPContentDecoder(ByteBufferPool pool, int bufferSize)
+    {
+        _bufferSize = bufferSize;
+        _pool = pool;
+        reset();
+    }
+
+    /** Inflate compressed data from a buffer.
+     * 
+     * @param compressed Buffer containing compressed data.
+     * @return Buffer containing inflated data.
+     */
+    public ByteBuffer decode(ByteBuffer compressed)
+    {
+        decodeChunks(compressed);
+        if (BufferUtil.isEmpty(_inflated) || _state==State.CRC || _state==State.ISIZE )
+            return BufferUtil.EMPTY_BUFFER;
+        
+        ByteBuffer result = _inflated;
+        _inflated = null;
+        return result;
+    }
+
+    /** Called when a chunk of data is inflated.
+     * <p>The default implementation aggregates all the chunks 
+     * into a single buffer returned from {@link #decode(ByteBuffer)}.
+     * Derived implementations may choose to consume chunks individually
+     * and return false to prevent further inflation until a subsequent
+     * call to {@link #decode(ByteBuffer)} or {@link #decodeChunks(ByteBuffer)}.
+     *  
+     * @param chunk The inflated chunk of data
+     * @return False if inflating should continue, or True if the call 
+     * to {@link #decodeChunks(ByteBuffer)} or {@link #decode(ByteBuffer)} 
+     * should return, allowing back pressure of compressed data.
+     */
+    protected boolean decodedChunk(ByteBuffer chunk)
+    {
+        if (_inflated==null)
+            _inflated=chunk;
+        else
+        {
+            int size = _inflated.remaining() + chunk.remaining();
+            if (size<=_inflated.capacity())
+            {
+                BufferUtil.append(_inflated,chunk);
+                BufferUtil.put(chunk,_inflated);
+                release(chunk);
+            }
+            else
+            {
+                ByteBuffer bigger=_pool==null?BufferUtil.allocate(size):_pool.acquire(size,false);
+                int pos=BufferUtil.flipToFill(bigger);
+                BufferUtil.put(_inflated,bigger);
+                BufferUtil.put(chunk,bigger);
+                BufferUtil.flipToFlush(bigger,pos);
+                release(_inflated);
+                release(chunk);
+                _inflated = bigger;
+            }
+        }
+        
+        return false;
+    }
+    
+    
+    /**
+     * Inflate compressed data.
+     * <p>Inflation continues until the compressed block end is reached, there is no 
+     * more compressed data or a call to {@link #decodedChunk(ByteBuffer)} returns true.
+     * @param compressed
+     */
+    protected void decodeChunks(ByteBuffer compressed)
+    {
+        try
+        {
+            while (true)
+            {
+                switch (_state)
```

Comment:
Split the state machine into states that consume byte by byte and those that do not.